File: ark_oauth_client_nodejs/examples/setup-check.js
/**
 * Checks this application's configuration against the provider's live metadata.
 *
 *   node examples/setup-check.js
 *
 * Worth running in CI, or rendering on a health page. Without it, the first symptom of a wrong
 * port, a stopped provider, a missing tenant id or a scope the client was never registered for is
 * `invalid_request` on a page a user is looking at.
 */
import { ArkOAuthClient } from '../src/index.js';
import { authority, clientId, clientSecret, origin } from './config.js';

const client = new ArkOAuthClient({
  authority,
  clientId,
  clientSecret,
  redirectUri: `${origin}/signin-oidc`,
  requireHttps: origin.startsWith('https')
});

const report = await client.checkSetup();

console.log(`authority     ${report.authority}`);
console.log(`client        ${report.clientId} (${report.tokenEndpointAuthMethod})`);
console.log(`redirect_uri  ${report.redirectUri}`);
console.log(`discovery     ${report.discoveryOk ? 'ok' : `FAILED — ${report.discoveryError}`}`);

if (report.discoveryOk) {
  console.log(`issuer        ${report.provider.issuer}`);
  console.log(`token         ${report.provider.tokenEndpoint}`);
  console.log(`keys          ${report.signingKeys?.map((k) => `${k.kid} (${k.kty})`).join(', ')}`);
  console.log(`scopes        ${report.provider.scopesSupported.join(' ')}`);
  console.log(`grants        ${report.provider.grantTypesSupported.join(' ')}`);
}

if (report.problems.length === 0) {
  console.log('\nNo problems found.');
} else {
  console.log(`\n${report.problems.length} problem(s):`);
  for (const problem of report.problems) console.log(`  • ${problem}`);
  process.exitCode = 1;
}

File: ark_oauth_client_py/examples/setup_check.py
"""
Checking the setup — run it in CI, or render it on a health page.

    ARK_AUTHORITY=... ARK_CLIENT_ID=... python examples/setup_check.py

Pairs local configuration with the provider's live metadata and reports what does not line up: a
wrong port, a stopped provider, an issuer mismatch, a scope this client is not registered for, an
authentication method the tenant does not offer. Without it, the first symptom of any of these is
`invalid_request` on a page a user is looking at.

Exits non-zero when something is wrong, so it fails a pipeline rather than a sign-in.
"""

import sys

from ark_oauth_client import ArkAuthConfig, ArkOAuthClient, ArkSetupProbe

from config import ORIGIN, REQUIRE_HTTPS, client_config


def main():
    config = ArkAuthConfig.from_mapping(client_config())

    # The renderable view: everything an operator needs to check a registration.
    model = ArkSetupProbe(config).probe(origin=ORIGIN)
    print(f"authority          : {model.authority}")
    print(f"client_id          : {model.client_id}")
    print(f"redirect_uri       : {model.redirect_uri}   <- register this exactly")
    print(f"post_logout_uri    : {model.post_logout_redirect_uri}")
    print(f"discovery          : {'ok' if model.discovery_ok else model.discovery_error}")

    if model.discovery_ok:
        print(f"issuer             : {model.issuer}")
        print(f"issuer mismatch    : {model.issuer_mismatch}")
        print(f"dynamic reg.       : {model.supports_dynamic_registration}")
        print(f"client credentials : {model.supports_client_credentials}")
        print(f"admin console      : {model.admin_console_url}")
        print(f"integration page   : {model.integration_page_url}")
        if model.unsupported_scopes:
            print(f"unpublished scopes : {', '.join(model.unsupported_scopes)}")

    # The protocol client's own report adds the problems as sentences.
    client = ArkOAuthClient(
        authority=config.resolve_authority(),
        client_id=config.client_id,
        client_secret=config.client_secret or None,
        require_https=REQUIRE_HTTPS,
    )
    report = client.check_setup(origin=ORIGIN)

    if report["problems"]:
        print("\nProblems:")
        for problem in report["problems"]:
            print(f"  - {problem}")
        sys.exit(1)

    print("\nNo problems found.")


if __name__ == "__main__":
    main()
